src/lib/getDataAndInitListeners.js

```diff
@@ -0,0 +1,169 @@
+const { ethers } = require('ethers');
+const { actions } = require('../store');
+const config = require('../config');
+const {
+  calculateCurrentTrancheId,
+  fetchStakingPoolsData,
+  fetchStakingPoolDataById,
+  fetchAllProductsData,
+  fetchProductDataById,
+  fetchStakingPoolIdAndAddress,
+  fetchProductDataForPool,
+  fetchAllProductDataForPool,
+} = require('../contract/helpers');
+const StakingPoolAbi = require('../abis/StakingPool.json');
+const StakingPoolFactoryAbi = require('../abis/StakingPoolFactory.json');
+const constants = require('./constants');
+const CoverAbi = require('../abis/Cover.json');
+
+const { CONTRACTS_ADDRESSES } = constants;
+
+const url = config.get('provider.ws');
+const wsProvider = new ethers.providers.WebSocketProvider(url);
+
+const stakingPoolContracts = [];
+let trancheId;
+let store;
+
+async function fetchInitialData() {
+  const stakingPools = await fetchStakingPoolsData();
+  const products = await fetchAllProductsData();
+  store.dispatch({
+    type: actions.SET_STATE,
+    payload: {
+      products,
+      stakingPools,
+    },
+  });
+  console.info('Initial Data set');
+}
+
+async function fetchStakingPoolData(id) {
+  const poolData = await fetchStakingPoolDataById(id);
+
+  store.dispatch({
+    type: actions.ADD_STAKING_POOL,
+    payload: { id, poolData },
+  });
+}
+
+async function updateProductsByStakingPool(poolId) {
+  const { products } = store.getState();
+  const stakingPoolProducts = Object.values(products).reduce((acc, [productId, stakingPools = []]) => {
+    if (Object.keys(stakingPools).includes(poolId)) {
+      acc.push(productId);
+    }
+    return acc;
+  }, []);
+  for (const productId of stakingPoolProducts) {
+    const stakingPoolData = fetchProductDataForPool(productId, poolId);
+    store.dispatch({
+      type: actions.SET_PRODUCT_STAKING_POOL,
+      payload: {
+        productId,
+        poolId,
+        stakingPoolData,
+      },
+    });
+  }
+}
+
+function subscribeToStakingPoolDependantEvents(poolId, address) {
+  const contract = new ethers.Contract(address, StakingPoolAbi, wsProvider);
+
+  contract.on('StakeBurned', () => updateProductsByStakingPool(poolId));
+  contract.on('DepositExtended', () => updateProductsByStakingPool(poolId));
+  contract.on('StakeDeposited', () => updateProductsByStakingPool(poolId));
+  contract.on('PoolFeeChanged', () => updateProductsByStakingPool(poolId));
+  contract.on('StakeBurned', () => updateProductsByStakingPool(poolId));
```

Comment:
Should be possible to combine in a single subscription, see `(topic[0] = A) OR (topic[0] = B)` https://docs.ethers.org/v5/single-page/#/v5/concepts/events/-%23-events--filters